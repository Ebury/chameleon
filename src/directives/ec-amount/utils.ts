
import defaultOptions from './options';
import type { AmountDirectiveOptions } from './types';

function format(input: undefined, opts?: AmountDirectiveOptions): undefined;
function format(input: null, opts?: AmountDirectiveOptions): null;
function format(input: string | number, opts?: AmountDirectiveOptions): string;
function format(input: string | number | undefined | null, opts: AmountDirectiveOptions = defaultOptions): string | null | undefined {
  /* eslint-disable no-param-reassign */
  if (input === '' || input === null || input === undefined) {
    return input;
  }

  if (typeof input === 'number') {
    input = input.toString();
  }

  const negative = input.startsWith('-') ? '-' : '';

  input = sanitizeInput(input, opts.decimalSeparator);
  if (opts.precision > 0) {
    input = keepOnlyOneDecimalSeparator(input, opts.decimalSeparator);
  } else {
    input = removeDecimal(input, opts.decimalSeparator);
  }
  const parts = input.split(opts.decimalSeparator);
  let integralPart = parseToStr(parts[0]);
  let fractionalPart = parts[1];

  if (fractionalPart && fractionalPart.length > opts.precision) {
    fractionalPart = fractionalPart.substring(0, opts.precision);
  }

  integralPart = addGroupingSeparators(integralPart, opts.groupingSeparator);
  const result = negative + joinIntegralAndFractionalParts(integralPart, fractionalPart, opts.decimalSeparator);
  return result;
}

function parseToStr(number: string): string {
  const parsed = Number.parseFloat(number);
  if (Number.isNaN(parsed)) {
    return '';
  }
  return parsed.toString();
}

function keepOnlyOneDecimalSeparator(number: string, decimalSeparator: string): string {
  if (number.includes(decimalSeparator)) {
    const split = number.split(decimalSeparator);
    return split[0] + decimalSeparator + split.slice(1).join('');
  }
  return number;
}

function removeDecimal(number: string, decimalSeparator: string): string {
  const split = number.split(decimalSeparator);
  return split[0];
}

function unFormat(input: string, groupingSeparator: string, decimalSeparator: string): string {
  if (!input) {
    return input;
  }

  return input
    .replace(new RegExp(`\\${groupingSeparator}`, 'g'), '')
    .replace(new RegExp(`\\${decimalSeparator}`, 'g'), '.');
}

function sanitizeInput(input: string, decimalSeparator: string): string {
  return input.replace(new RegExp(`[^0-9${decimalSeparator}]`, 'g'), '') || '';
}

function addGroupingSeparators(integralPart: string, separator: string): string {
  return integralPart.replace(/(\d)(?=(?:\d{3})+\b)/gm, `$1${separator}`);
}

function joinIntegralAndFractionalParts(integralPart: string, fractionalPart: string, decimalSeparator: string): string {
  return fractionalPart !== undefined ? integralPart + decimalSeparator + fractionalPart : integralPart;
}

/* c8 ignore start */
function setCursor(el: HTMLInputElement, position: number) {
  const setSelectionRange = function setSelectionRange() {
    el.setSelectionRange(position, position);
  };

  if (el === document.activeElement) {
    setSelectionRange();
    setTimeout(setSelectionRange, 1); // Android Fix
  }
}
/* c8 ignore stop */

export {
  format,
  sanitizeInput,
  setCursor,
  unFormat,
};
