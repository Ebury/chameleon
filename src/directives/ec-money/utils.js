import defaults from './options';

function format(input, opt = defaults) {
  if (input === '' || input === null || input === undefined) {
    return input;
  }

  if (typeof input === 'number') {
    input = input.toString();
  }

  const negative = input.includes('-') ? '-' : '';

  input = sanitizeInput(input, opt.decimalSeparator);
  input = keepOnlyOneDecimalSeparator(input, opt.decimalSeparator);
  const parts = input.split(opt.decimalSeparator);
  let integer = parseToStr(parts[0]);
  let decimal = parts[1];

  if (decimal && decimal.length > opt.precision) {
    decimal = decimal.substr(0, opt.precision);
  }

  integer = addGroupingSeparators(integer, opt.groupingSeparator);
  const result = negative + joinIntegerAndDecimal(integer, decimal, opt.decimalSeparator);
  return result;
}

function parseToStr(number) {
  /* eslint no-restricted-globals: "off" */
  const parsed = parseFloat(number);
  if (isNaN(parsed)) {
    return '';
  }
  return parsed.toString();
}

function keepOnlyOneDecimalSeparator(number, decimalSeparator) {
  if (number.includes(decimalSeparator)) {
    const split = number.split(decimalSeparator);
    return split[0] + decimalSeparator + split.slice(1).join('');
  }
  return number;
}

function unFormat(input, groupingSeparator, decimalSeparator) {
  /* eslint no-param-reassign: "off" */
  input = (`${input}`).replace(new RegExp(`\\${groupingSeparator}`, 'g'), '');
  input = (`${input}`).replace(new RegExp(`\\${decimalSeparator}`, 'g'), '.');
  return input;
}

function sanitizeInput(input, decimalSeparator) {
  return input.replace(new RegExp(`[^0-9${decimalSeparator}]`, 'g'), '') || '';
}

function addGroupingSeparators(integer, separator) {
  return integer.replace(/(\d)(?=(?:\d{3})+\b)/gm, `$1${separator}`);
}

function joinIntegerAndDecimal(integer, decimal, decimalSeparator) {
  return decimal !== undefined ? integer + decimalSeparator + decimal : integer;
}

function setCursor(el, position) {
  const setSelectionRange = function setSelectionRange() {
    el.setSelectionRange(position, position);
  };

  if (el === document.activeElement) {
    setSelectionRange();
    setTimeout(setSelectionRange, 1); // Android Fix
  }
}

// https://developer.mozilla.org/en-US/docs/Web/Guide/Events/Creating_and_triggering_events#The_old-fashioned_way
function event(name) {
  const evt = document.createEvent('Event');
  evt.initEvent(name, true, true);
  return evt;
}

export {
  format,
  unFormat,
  setCursor,
  event,
  sanitizeInput,
};
