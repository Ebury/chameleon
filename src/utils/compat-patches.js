/* istanbul ignore file */

import {
  Dropdown as FvDropdown,
  TooltipDirective as FvTooltipDirective,
  VClosePopper,
  VTooltip,
} from 'floating-vue';

// we need to tell the @vue/compat that all components coming from floating-vue are Vue 3 components so it skips
// them when patching component instances. These components can't run in default mode for Vue 2, there is lots of
// errors coming from the console in that mode.
export default function applyCompatPatches() {
  const compatConfig = {
    MODE: 3,
    RENDER_FUNCTION: false,
  };

  VTooltip.compatConfig = compatConfig;
  VClosePopper.compatConfig = compatConfig;
  FvTooltipDirective.compatConfig = compatConfig;
  for (const component of Object.values(FvTooltipDirective.components)) {
    component.compatConfig = compatConfig;
  }

  FvDropdown.compatConfig = compatConfig;

  for (const component of Object.values(FvDropdown.components)) {
    component.compatConfig = compatConfig;
  }
}
