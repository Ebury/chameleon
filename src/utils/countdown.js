import mitt from 'mitt';

export default class Countdown {
  constructor() {
    const m = mitt();
    Object.assign(this, m);
  }

  reduceMinutesLeft(totalSeconds) {
    this.getminutes(totalSeconds);
    this.getseconds(totalSeconds, this.minutesLeft);
  }

  start(secondsToGo, showMinutes) {
    this.seconds = secondsToGo;
    this.secondsLeft = secondsToGo;
    this.currentTime = Math.ceil(Date.now() / 1000);
    this.startTime = Math.ceil(Date.now() / 1000);
    this.interval = setInterval(() => this.reduceSecondsLeft(showMinutes), 1000);
  }

  get timeDifference() {
    return this.currentTime - this.startTime;
  }

  reduceSecondsLeft(showMinutes) {
    this.currentTime = Math.ceil(Date.now() / 1000);
    this.totalSeconds = this.seconds - this.timeDifference;
    this.minutesLeft = null;
    this.secondsLeft = this.totalSeconds;

    if (showMinutes) {
      this.reduceMinutesLeft(this.totalSeconds);
    }

    this.emit('time-updated', {
      seconds: this.secondsLeft,
      minutes: this.minutesLeft,
      totalSeconds: this.totalSeconds,
    });

    if (this.totalSeconds <= 0) {
      this.stop();
      this.emit('time-expired');
    }
  }

  stop() {
    clearInterval(this.interval);
  }

  getminutes(totalSeconds) {
    this.minutesLeft = Math.floor(totalSeconds / 60);
  }

  getseconds(totalSeconds, minutes) {
    // take mins remaining (as seconds) away from total seconds remaining
    this.secondsLeft = totalSeconds - Math.round(minutes * 60);
  }
}
