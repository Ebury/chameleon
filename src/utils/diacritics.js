export function removeDiacritics(str) {
  if (typeof str === 'string' && str.normalize) {
    // https://stackoverflow.com/questions/990904/remove-accents-diacritics-in-a-string-in-javascript
    return str.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
  }

  // IE11 users will not enjoy this feature:
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize
  return str;
}
