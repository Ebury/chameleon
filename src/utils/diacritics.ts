export function removeDiacritics(str: string): string {
  if (str && str.normalize) {
    // https://stackoverflow.com/questions/990904/remove-accents-diacritics-in-a-string-in-javascript
    return str.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
  }

  // IE11 users will not enjoy this feature:
  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/normalize
  return str;
}
