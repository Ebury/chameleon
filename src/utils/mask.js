export function mask(text, maskSymbol = '*', visibleChars = 2) {
  if (!text) {
    throw new Error('Text is required');
  }
  if (!maskSymbol) {
    throw new Error('Mask symbol cannot be empty');
  }
  if (!visibleChars || visibleChars < 0) {
    throw new Error('Visible characters must be a number greater than zero');
  }

  const digitsToBeMasked = text.length - visibleChars;
  return maskSymbol.repeat(digitsToBeMasked) + text.substr(digitsToBeMasked);
}
