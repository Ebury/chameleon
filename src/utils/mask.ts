export function mask(text: string, maskSymbol: string = '*', visibleChars: number = 2): string {
  if (!text) {
    throw new Error('Text is required');
  }
  if (!maskSymbol) {
    throw new Error('Mask symbol cannot be empty');
  }
  if (!visibleChars || visibleChars < 0) {
    throw new Error('Visible characters must be a number greater than zero');
  }

  const digitsToBeMasked = text.length - visibleChars;
  return maskSymbol.repeat(digitsToBeMasked) + text.substring(digitsToBeMasked);
}
