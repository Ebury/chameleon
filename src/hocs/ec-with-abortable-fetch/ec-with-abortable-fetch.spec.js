import { flushPromises, mount } from '@vue/test-utils';
import { defineComponent, h } from 'vue';

import { withMockedConsole } from '../../../tests/utils/console';
import withAbortableFetch from './ec-with-abortable-fetch';

describe('EcWithAbortableFetch', () => {
  function mountEcWithAbortableFetch(props, mountOpts, customProps) {
    const Component = defineComponent({
      name: 'WrappedComponent',
      props: [customProps?.dataProp ?? 'data', customProps?.errorProp ?? 'error', customProps?.loadingProp ?? 'loading'],
      render() {
        return h('div');
      },
    });

    const hocWrapper = mount(withAbortableFetch(Component, customProps), {
      props,
      ...mountOpts,
    });

    const componentWrapper = hocWrapper.findComponent(Component);

    return { hocWrapper, componentWrapper };
  }

  it('should throw an error if dataSource prop is missing', () => {
    withMockedConsole((errorSpy, warnSpy) => {
      mountEcWithAbortableFetch();
      expect(warnSpy).toHaveBeenCalledTimes(1);
      expect(warnSpy.mock.calls[0][0]).toContain('Missing required prop: "dataSource"');
    });
  });

  it('should throw an error if dataSource.fetch function is missing', () => {
    withMockedConsole((errorSpy, warnSpy) => {
      mountEcWithAbortableFetch({ dataSource: {} });
      expect(warnSpy).toHaveBeenCalledTimes(1);
      expect(warnSpy.mock.calls[0][0]).toContain('Invalid prop: custom validator check failed for prop "dataSource".');
    });
  });

  it('should throw an error if dataSource.fetch is not a function', () => {
    withMockedConsole((errorSpy, warnSpy) => {
      mountEcWithAbortableFetch({ dataSource: { fetch: true } });
      expect(warnSpy).toHaveBeenCalledTimes(1);
      expect(warnSpy.mock.calls[0][0]).toContain('Invalid prop: custom validator check failed for prop "dataSource".');
    });
  });

  it('should start fetching when created', () => {
    const dataSource = {
      fetch: jest.fn(),
    };

    const { componentWrapper } = mountEcWithAbortableFetch({ dataSource });
    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot();
    expect(dataSource.fetch).toMatchSnapshot();
  });

  it('should stop fetching and pass the data to the component after dataSource gets resolved', async () => {
    const data = { prop: 1 };
    const dataSource = {
      fetch: jest.fn().mockResolvedValueOnce(data),
    };

    const { componentWrapper } = mountEcWithAbortableFetch({ dataSource });
    await flushPromises();
    expect(getWrappedComponentState(componentWrapper).data).toEqual(data);
    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot();
    expect(dataSource.fetch).toMatchSnapshot();
  });

  it('should stop fetching and pass the error to the component after dataSource gets rejected', async () => {
    const error = new Error('Random error');
    const dataSource = {
      fetch: jest.fn().mockRejectedValueOnce(error),
    };

    const errorSpy = jest.fn();

    const { componentWrapper } = mountEcWithAbortableFetch({ dataSource }, {
      attrs: {
        onError: errorSpy,
      },
    });
    await flushPromises();
    expect(getWrappedComponentState(componentWrapper).error).toBe(error);
    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot();
    expect(dataSource.fetch).toMatchSnapshot();
    expect(errorSpy.mock.calls).toMatchSnapshot();
  });

  it('should not propagate the error when re-fetching after an error occurred', async () => {
    const error = new Error('Random error');
    const dataSource = {
      fetch: jest.fn().mockRejectedValueOnce(error).mockResolvedValueOnce({ result: 1 }),
    };

    const errorSpy = jest.fn();

    const { componentWrapper, hocWrapper } = mountEcWithAbortableFetch({ dataSource }, {
      attrs: {
        onError: errorSpy,
      },
    });
    await flushPromises();
    expect(getWrappedComponentState(componentWrapper).error).toBe(error);
    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot();
    expect(dataSource.fetch).toMatchSnapshot();
    expect(errorSpy.mock.calls).toMatchSnapshot();

    dataSource.fetch.mockClear();
    errorSpy.mockClear();

    await hocWrapper.setProps({
      fetchArgs: { anotherProp: 2 },
    });

    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot();
    expect(dataSource.fetch).toMatchSnapshot();

    await flushPromises();
    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot();
    expect(getWrappedComponentState(componentWrapper).error).toBe(null);
    expect(errorSpy).not.toHaveBeenCalled();
  });

  it('should send the given fetch arguments', () => {
    const dataSource = {
      fetch: jest.fn(),
    };
    const fetchArgs = { prop: 1 };

    const { componentWrapper } = mountEcWithAbortableFetch({ dataSource, fetchArgs });
    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot();
    expect(dataSource.fetch).toMatchSnapshot();
  });

  it('should re-fetch the data if arguments change', async () => {
    const dataSource = {
      fetch: jest.fn().mockResolvedValueOnce({ result: 1 }).mockResolvedValueOnce({ result: 2 }),
    };
    const fetchArgs = { prop: 1 };

    const { componentWrapper, hocWrapper } = mountEcWithAbortableFetch({ dataSource, fetchArgs });
    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot();
    expect(dataSource.fetch).toMatchSnapshot();

    await flushPromises();
    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot();

    dataSource.fetch.mockClear();

    await hocWrapper.setProps({
      fetchArgs: { anotherProp: 2 },
    });

    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot();
    expect(dataSource.fetch).toMatchSnapshot();

    await flushPromises();
    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot();
  });

  it('should abort previous fetch calls', async () => {
    const dataSource = {
      fetch: jest.fn()
        .mockImplementationOnce(() => new Promise((resolve) => { setTimeout(() => resolve({ result: 1 }), 1); }))
        .mockImplementationOnce(() => new Promise((resolve) => { setTimeout(() => resolve({ result: 2 }), 1); })),
    };
    const fetchArgs = { prop: 1 };
    const abortSpy = jest.fn();

    const { componentWrapper, hocWrapper } = mountEcWithAbortableFetch({ dataSource, fetchArgs }, {
      attrs: {
        onAbort: abortSpy,
      },
    });
    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot();

    const cancelToken = dataSource.fetch.mock.calls[0][1];
    expect(cancelToken).toBeInstanceOf(global.AbortSignal);
    cancelToken.addEventListener('abort', abortSpy);

    hocWrapper.setProps({
      fetchArgs: { anotherProp: 2 },
    });

    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot();
    await flushPromises();
    await flushPromises(); // because of two mockImplementationOnce of the fetch
    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot();
    expect(abortSpy).toHaveBeenCalledTimes(2);
  });

  it('should abort fetch call when destroyed', () => {
    const dataSource = {
      fetch: jest.fn(),
    };
    const abortSpy = jest.fn();

    const { componentWrapper, hocWrapper } = mountEcWithAbortableFetch({ dataSource }, {
      attrs: {
        onAbort: abortSpy,
      },
    });

    const cancelToken = dataSource.fetch.mock.calls[0][1];
    expect(cancelToken).toBeInstanceOf(global.AbortSignal);
    cancelToken.addEventListener('abort', abortSpy);

    hocWrapper.unmount();

    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot();
    expect(abortSpy).toHaveBeenCalledTimes(2);
  });

  it('should stop fetching and keep the state intact if fetch for rejected because of AbortError', async () => {
    class AbortError extends Error {// AbortError is not exposed in the DOM
      constructor() {
        super();
        this.name = 'AbortError';
      }
    }
    const error = new AbortError();
    const dataSource = {
      fetch: jest.fn().mockRejectedValueOnce(error),
    };

    const errorSpy = jest.fn();

    const { componentWrapper } = mountEcWithAbortableFetch({ dataSource }, {
      attrs: {
        onError: errorSpy,
      },
    });
    await flushPromises();
    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot();
    expect(dataSource.fetch).toHaveBeenCalledTimes(1);
    expect(errorSpy).not.toHaveBeenCalled();
  });

  it('should map the given custom props', async () => {
    const dataSource = {
      fetch: jest.fn().mockResolvedValueOnce({ result: 1 }),
    };
    const fetchArgs = { prop: 1 };
    const customProps = {
      dataProp: 'dataCustom',
      loadingProp: 'isLoadingCustom',
      errorProp: 'errorCustom',
    };
    const { componentWrapper } = mountEcWithAbortableFetch({ dataSource, fetchArgs }, {}, customProps);
    expect(getWrappedComponentState(componentWrapper, customProps)).toMatchSnapshot('while loading');
    await flushPromises();
    expect(getWrappedComponentState(componentWrapper, customProps)).toMatchSnapshot('after fetch');
  });

  it('should transform props using given data transform function', async () => {
    const dataSource = {
      fetch: jest.fn().mockResolvedValueOnce({ result: 1 }),
    };

    const dataTransform = jest.fn().mockImplementation(data => data ?? { result: 0 });

    const customProps = {
      dataTransform,
    };
    const { componentWrapper } = mountEcWithAbortableFetch({ dataSource }, {}, customProps);
    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot('while loading');
    await flushPromises();
    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot('after fetch');
    expect(dataTransform).toHaveBeenCalledTimes(2);
    expect(dataTransform).toHaveBeenNthCalledWith(1, null);
    expect(dataTransform).toHaveBeenNthCalledWith(2, { result: 1 });
  });

  it('should transform props using given data transform function and pass them to given custom data prop', async () => {
    const dataSource = {
      fetch: jest.fn().mockResolvedValueOnce({ result: 1 }),
    };

    const dataTransform = jest.fn().mockImplementation(data => data ?? { result: 0 });

    const customProps = {
      dataTransform,
      dataProp: 'dataCustom',
    };
    const { componentWrapper } = mountEcWithAbortableFetch({ dataSource }, {}, customProps);
    expect(getWrappedComponentState(componentWrapper, customProps)).toMatchSnapshot('while loading');
    await flushPromises();
    expect(getWrappedComponentState(componentWrapper, customProps)).toMatchSnapshot('after fetch');
    expect(dataTransform).toHaveBeenCalledTimes(2);
    expect(dataTransform).toHaveBeenNthCalledWith(1, null);
    expect(dataTransform).toHaveBeenNthCalledWith(2, { result: 1 });
  });

  it('should transform props using given error transform function', async () => {
    const error = new Error('Random error');
    const dataSource = {
      fetch: jest.fn().mockRejectedValueOnce(error),
    };

    const errorTransform = jest.fn().mockImplementation((err) => {
      if (err) {
        return `Custom error message (${err.message})`;
      }
      return null;
    });

    const customProps = {
      errorTransform,
    };
    const { componentWrapper } = mountEcWithAbortableFetch({ dataSource }, {}, customProps);
    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot('while loading');
    await flushPromises();
    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot('after fetch');
    expect(errorTransform).toHaveBeenCalledTimes(2);
    expect(errorTransform).toHaveBeenNthCalledWith(1, null);
    expect(errorTransform).toHaveBeenNthCalledWith(2, error);
  });

  it('should transform props using given error transform function and pass them to given custom error prop', async () => {
    const error = new Error('Random error');
    const dataSource = {
      fetch: jest.fn().mockRejectedValueOnce(error),
    };

    const errorTransform = jest.fn().mockImplementation((err) => {
      if (err) {
        return `Custom error message (${err.message})`;
      }
      return null;
    });

    const customProps = {
      errorTransform,
      errorProp: 'errorCustom',
    };
    const { componentWrapper } = mountEcWithAbortableFetch({ dataSource }, {}, customProps);
    expect(getWrappedComponentState(componentWrapper, customProps)).toMatchSnapshot('while loading');
    await flushPromises();
    expect(getWrappedComponentState(componentWrapper, customProps)).toMatchSnapshot('after fetch');
    expect(errorTransform).toHaveBeenCalledTimes(2);
    expect(errorTransform).toHaveBeenNthCalledWith(1, null);
    expect(errorTransform).toHaveBeenNthCalledWith(2, error);
  });

  it('should transform props using given loading transform function', async () => {
    const dataSource = {
      fetch: jest.fn().mockResolvedValueOnce({ result: 1 }),
    };

    const loadingTransform = jest.fn().mockImplementation(isLoading => !isLoading);

    const customProps = {
      loadingTransform,
    };
    const { componentWrapper } = mountEcWithAbortableFetch({ dataSource }, {}, customProps);
    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot('while loading');
    await flushPromises();
    expect(getWrappedComponentState(componentWrapper)).toMatchSnapshot('after fetch');
    expect(loadingTransform).toHaveBeenCalledTimes(2);
    expect(loadingTransform).toHaveBeenNthCalledWith(1, true, null);
    expect(loadingTransform).toHaveBeenNthCalledWith(2, false, { result: 1 });
  });

  it('should transform props using given loading transform function and pass them to given custom loading prop and custom data prop', async () => {
    const dataSource = {
      fetch: jest.fn().mockResolvedValueOnce({ result: 1 }),
    };

    const loadingTransform = jest.fn().mockImplementation(isLoading => !isLoading);

    const customProps = {
      loadingTransform,
      loadingProp: 'loadingCustom',
      dataProp: 'dataCustom',
    };
    const { componentWrapper } = mountEcWithAbortableFetch({ dataSource }, {}, customProps);
    expect(getWrappedComponentState(componentWrapper, customProps)).toMatchSnapshot('while loading');
    await flushPromises();
    expect(getWrappedComponentState(componentWrapper, customProps)).toMatchSnapshot('after fetch');
    expect(loadingTransform).toHaveBeenCalledTimes(2);
    expect(loadingTransform).toHaveBeenNthCalledWith(1, true, null);
    expect(loadingTransform).toHaveBeenNthCalledWith(2, false, { result: 1 });
  });

  it('should stop fetching and map the custom error prop to the component after dataSource gets rejected', async () => {
    const error = new Error('Random error');
    const dataSource = {
      fetch: jest.fn().mockRejectedValueOnce(error),
    };

    const errorSpy = jest.fn();

    const customProps = {
      dataProp: 'dataCustom',
      loadingProp: 'isLoadingCustom',
      errorProp: 'errorCustom',
    };

    const { componentWrapper } = mountEcWithAbortableFetch({ dataSource }, {
      attrs: {
        onError: errorSpy,
      },
    }, customProps);
    await flushPromises();
    expect(getWrappedComponentState(componentWrapper, customProps)[customProps.errorProp]).toBe(error);
    expect(getWrappedComponentState(componentWrapper, customProps)).toMatchSnapshot();
    expect(dataSource.fetch).toMatchSnapshot();
    expect(errorSpy.mock.calls).toMatchSnapshot();
  });
});

function getWrappedComponentState(componentWrapper, { dataProp = 'data', loadingProp = 'loading', errorProp = 'error' } = {}) {
  const { [loadingProp]: loading, [dataProp]: data, [errorProp]: error } = componentWrapper.vm;
  return { [loadingProp]: loading, [dataProp]: data, [errorProp]: error };
}
