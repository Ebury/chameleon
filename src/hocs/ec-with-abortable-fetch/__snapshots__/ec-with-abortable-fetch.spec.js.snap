// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EcWithAbortableFetch should abort fetch call when destroyed 1`] = `
Object {
  "data": null,
  "error": null,
  "loading": true,
}
`;

exports[`EcWithAbortableFetch should abort previous fetch calls 1`] = `
Object {
  "data": null,
  "error": null,
  "loading": true,
}
`;

exports[`EcWithAbortableFetch should abort previous fetch calls 2`] = `
Object {
  "data": null,
  "error": null,
  "loading": true,
}
`;

exports[`EcWithAbortableFetch should abort previous fetch calls 3`] = `
Object {
  "data": Object {
    "result": 2,
  },
  "error": null,
  "loading": false,
}
`;

exports[`EcWithAbortableFetch should not propagate the error when re-fetching after an error occurred 1`] = `
Object {
  "data": null,
  "error": [Error: Random error],
  "loading": false,
}
`;

exports[`EcWithAbortableFetch should not propagate the error when re-fetching after an error occurred 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      undefined,
      AbortSignal {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`EcWithAbortableFetch should not propagate the error when re-fetching after an error occurred 3`] = `
Array [
  Array [
    [Error: Random error],
  ],
]
`;

exports[`EcWithAbortableFetch should not propagate the error when re-fetching after an error occurred 4`] = `
Object {
  "data": null,
  "error": null,
  "loading": true,
}
`;

exports[`EcWithAbortableFetch should not propagate the error when re-fetching after an error occurred 5`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "anotherProp": 2,
      },
      AbortSignal {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`EcWithAbortableFetch should not propagate the error when re-fetching after an error occurred 6`] = `
Object {
  "data": Object {
    "result": 1,
  },
  "error": null,
  "loading": false,
}
`;

exports[`EcWithAbortableFetch should re-fetch the data if arguments change 1`] = `
Object {
  "data": null,
  "error": null,
  "loading": true,
}
`;

exports[`EcWithAbortableFetch should re-fetch the data if arguments change 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "prop": 1,
      },
      AbortSignal {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`EcWithAbortableFetch should re-fetch the data if arguments change 3`] = `
Object {
  "data": Object {
    "result": 1,
  },
  "error": null,
  "loading": false,
}
`;

exports[`EcWithAbortableFetch should re-fetch the data if arguments change 4`] = `
Object {
  "data": Object {
    "result": 1,
  },
  "error": null,
  "loading": true,
}
`;

exports[`EcWithAbortableFetch should re-fetch the data if arguments change 5`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "anotherProp": 2,
      },
      AbortSignal {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`EcWithAbortableFetch should re-fetch the data if arguments change 6`] = `
Object {
  "data": Object {
    "result": 2,
  },
  "error": null,
  "loading": false,
}
`;

exports[`EcWithAbortableFetch should send the given fetch arguments 1`] = `
Object {
  "data": null,
  "error": null,
  "loading": true,
}
`;

exports[`EcWithAbortableFetch should send the given fetch arguments 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "prop": 1,
      },
      AbortSignal {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`EcWithAbortableFetch should start fetching when created 1`] = `
Object {
  "data": null,
  "error": null,
  "loading": true,
}
`;

exports[`EcWithAbortableFetch should start fetching when created 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      undefined,
      AbortSignal {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`EcWithAbortableFetch should stop fetching and keep the state intact if fetch for rejected because of AbortError 1`] = `
Object {
  "data": null,
  "error": null,
  "loading": false,
}
`;

exports[`EcWithAbortableFetch should stop fetching and pass the data to the component after dataSource gets resolved 1`] = `
Object {
  "data": Object {
    "prop": 1,
  },
  "error": null,
  "loading": false,
}
`;

exports[`EcWithAbortableFetch should stop fetching and pass the data to the component after dataSource gets resolved 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      undefined,
      AbortSignal {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`EcWithAbortableFetch should stop fetching and pass the error to the component after dataSource gets rejected 1`] = `
Object {
  "data": null,
  "error": [Error: Random error],
  "loading": false,
}
`;

exports[`EcWithAbortableFetch should stop fetching and pass the error to the component after dataSource gets rejected 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      undefined,
      AbortSignal {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Promise {},
    },
  ],
}
`;

exports[`EcWithAbortableFetch should stop fetching and pass the error to the component after dataSource gets rejected 3`] = `
Array [
  Array [
    [Error: Random error],
  ],
]
`;

exports[`EcWithAbortableFetch should throw an error if dataSource prop is missing 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[Vue warn]: Missing required prop: \\"dataSource\\"

found in

---> <EcWithAbortableFetch>
       <Root>",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`EcWithAbortableFetch should throw an error if dataSource.fetch function is missing 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[Vue warn]: Invalid prop: custom validator check failed for prop \\"dataSource\\".

found in

---> <EcWithAbortableFetch>
       <Root>",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`EcWithAbortableFetch should throw an error if dataSource.fetch is not a function 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "[Vue warn]: Invalid prop: custom validator check failed for prop \\"dataSource\\".

found in

---> <EcWithAbortableFetch>
       <Root>",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
